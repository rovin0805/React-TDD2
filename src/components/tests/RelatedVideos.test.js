import {
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from '@testing-library/react';
import { Route } from 'react-router-dom';
import { withAllContexts, withRouter } from '../../tests/utils';
import { fakeVideos } from '../../tests/video';
import RelatedVideos from '../RelatedVideos';

describe('Related Videos', () => {
  const fakeYoutube = {
    relatedVideos: jest.fn(),
  };

  function renderRelatedVideosWithCallback(callback) {
    fakeYoutube.relatedVideos.mockImplementation(callback);
    return render(
      withAllContexts(
        withRouter(<Route path='/' element={<RelatedVideos id='id' />} />),
        fakeYoutube
      )
    );
  }

  afterEach(() => fakeYoutube.relatedVideos.mockReset());

  it('renders correctly', async () => {
    const { asFragment } = renderRelatedVideosWithCallback(() => fakeVideos);

    await waitForElementToBeRemoved(() => screen.queryByText('Loading...'));

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders related video correctly', async () => {
    renderRelatedVideosWithCallback(() => fakeVideos);

    expect(fakeYoutube.relatedVideos).toHaveBeenCalledWith('id');

    await waitFor(() =>
      expect(screen.getAllByRole('listitem')).toHaveLength(fakeVideos.length)
    );
  });

  it('renders loading', () => {
    renderRelatedVideosWithCallback(() => fakeVideos);

    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });

  it('renders error', async () => {
    renderRelatedVideosWithCallback(() => {
      throw new Error('error');
    });

    const errMsg = await screen.findByText('Something is wrong ðŸ˜–');
    expect(errMsg).toBeInTheDocument();
  });
});
