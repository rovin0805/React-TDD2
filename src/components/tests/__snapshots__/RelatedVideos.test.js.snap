// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Related Videos renders correctly 1`] = `
<DocumentFragment>
  <ul>
    <li
      class="flex gap-1 m-2"
    >
      <img
        alt="title"
        class="w-60 mr-2"
        src="http://localhost/"
      />
      <div>
        <p
          class="font-semibold my-2 line-clamp-2"
        >
          title
        </p>
        <p
          class="text-sm opacity-80"
        >
          1
        </p>
        <p
          class="text-sm opacity-80"
        >
          just now
        </p>
      </div>
    </li>
    <li
      class="flex gap-1 m-2"
    >
      <img
        alt="title2"
        class="w-60 mr-2"
        src="http://localhost/"
      />
      <div>
        <p
          class="font-semibold my-2 line-clamp-2"
        >
          title2
        </p>
        <p
          class="text-sm opacity-80"
        >
          2
        </p>
        <p
          class="text-sm opacity-80"
        >
          just now
        </p>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
